File: Workspace/week09/JSW/도전 문제/9663/solution_01.cpp
// Problem: N-Queen(https://www.acmicpc.net/problem/9663)
// Date: 2024.11.25
// Comment: 통과.
// -> 무적의 C++

#include <iostream>

static int N;
static int* state;
static bool* visited;
static int count;
bool isValid(int* state, int depth) {
    int y_up = state[depth-1];
    int y_down = state[depth-1];
    for (int i = depth-2; i >= 0; i--) {
        y_up += 1;
        y_down -= 1;
        if (state[i] == y_up || state[i] == y_down) {
            return false;
        }
    }
    return true;
}
void recurse(int depth) {
    if (!isValid(state, depth)) {
        return;
    }
    if (depth == N) {
        count++;
        return;
    }
    for (int i = 0; i < N; i++) {
        if (!visited[i]) {
            visited[i] = true;
            state[depth] = i;
            recurse(depth+1);
            visited[i] = false;
        }
    }
}
int main() {
    std::cin >> N;
    state = new int[N];
    visited = new bool[N];
    count = 0;
    recurse(0);
    std::cout << count;


    delete[] state;
    delete[] visited;
    
    return 0;
}

File: Workspace/week09/JSW/도전 문제/9663/solution_01.py
# Problem: N-Queen(https://www.acmicpc.net/problem/9663)
# Date: 2024.11.25
# Comment: 
# 적절하게 state space를 구성하고 백트래킹을 사용하여 코드를 작성하였다.
# -> 시간초과
# -> 일단 코드가 맞는지 확인하기 위해 N이 [1,15)일 때의 결과값을 배열에 저장하여 제출해보자.

def solution(N):
    answer = [0]*N
    visited = [False]*N
    history = []
    def is_valid(state, depth):
        y_0 = y_1 = state[depth-1]
        for i in range(depth-2, -1, -1):
            y_0, y_1 = y_0-1, y_1+1
            if y_0 == state[i] or y_1 == state[i]: return False
        return True
    def recurse(depth):
        if not is_valid(answer, depth):
            return
        if depth == N:
            history.append(answer[:])
            return
        for i in range(N):
            if visited[i]: continue
            answer[depth] = i
            visited[i] = True
            recurse(depth+1)
            visited[i] = False
    recurse(0)
    return len(history)

N = int(input())
print(solution(N))
